contrib/nydusify/pkg/chunkdict/generater/generater.go

```diff
@@ -0,0 +1,117 @@
+package generater
+
+import (
+	"context"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/dragonflyoss/image-service/contrib/nydusify/pkg/parser"
+	"github.com/dragonflyoss/image-service/contrib/nydusify/pkg/provider"
+	"github.com/dragonflyoss/image-service/contrib/nydusify/pkg/utils"
+)
+
+// Opt defines Chunkdict generate options.
+// Note: sources is one or more Nydus image references.
+type Opt struct {
+	WorkDir        string
+	Sources        string
+	SourceInsecure bool
+	NydusImagePath string
+	ExpectedArch   string
+}
+
+// "Chunkdict generate" command store multiple Nydus images into database
+// by invoking "nydus-image chunkdict save" command
+// to save chunk and blob information image into database.
+type Generater struct {
+	Opt
+	// Nydus image references slice.
+	sourcesSlice  []string
+	sourcesParser []*parser.Parser
+}
+
+// New creates Generater instance, sourceSlice is multiple Nydus images reference list.
+func New(opt Opt) (*Generater, error) {
+	// TODO: support sources resolver
+	sourcesSlice := strings.Split(opt.Sources, ",")
+	var sourcesParser []*parser.Parser
+	for _, source := range sourcesSlice {
+		sourcesRemote, err := provider.DefaultRemote(source, opt.SourceInsecure)
+		if err != nil {
+			return nil, errors.Wrap(err, "Init source image parser")
+		}
+		sourceParser, err := parser.New(sourcesRemote, opt.ExpectedArch)
+		sourcesParser = append(sourcesParser, sourceParser)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create parser")
+		}
+	}
+
+	generater := &Generater{
+		Opt:           opt,
+		sourcesSlice:  sourcesSlice,
+		sourcesParser: sourcesParser,
+	}
+
+	return generater, nil
+}
+
+// "Chunkdict Generate" deposit multiple Nydus images into the database one by one.
+func (generater *Generater) Generate(ctx context.Context) error {
+	for index := 0; index < len(generater.sourcesParser); index++ {
+		if err := generater.generate(ctx, index); err != nil {
+			if utils.RetryWithHTTP(err) {
+				generater.sourcesParser[index].Remote.MaybeWithHTTP(err)
+			}
+		}
+		generater.generate(ctx, index)
+	}
+	return nil
+}
+
+// A Nydus image is stored in the database
+func (generater *Generater) generate(ctx context.Context, index int) error {
+	sourceParsed, err := generater.sourcesParser[index].Parse(ctx)
+	if err != nil {
+		return errors.Wrap(err, "parse Nydus image")
+	}
+
+	// Create a directory to store the image bootstrap
+	nydusImageName := generater.sourcesSlice[index][strings.LastIndex(generater.sourcesSlice[index], "/")+1:]
+	folderPath := filepath.Join(generater.WorkDir, nydusImageName)
+	_, err = os.Stat(folderPath)
+	if err == nil {
```

Comment:
Can we use os.MkdirAll here?
`os.MkdirAll`:  `If path is already a directory, MkdirAll does nothing and returns nil.`

Reply:
And we need to clean up the work directory after the generation finish.